File: lib/08.js
function parseProgram(contents) {
	return contents.split(/\n/g).map((line) => {
		const [cmd, param] = line.split(' ');
		return {
			cmd,
			param: parseInt(param, 10),
			count: 0,
		};
	});
}

function execProgram(cmds) {
	cmds.forEach((item) => {
		item.count = 0;
	});

	let acc = 0;
	let pos = 0;
	let curr = cmds[pos];
	while (curr && curr.count < 1) {
		curr.count++;
		if (curr.cmd === 'jmp') {
			pos += curr.param;
		} else {
			if (curr.cmd === 'acc') {
				acc += curr.param;
			}
			pos++;
		}
		curr = cmds[pos];
	}

	return { acc, loop: Boolean(curr) };
}

module.exports = {
	A08(contents) {
		const cmds = parseProgram(contents);

		return execProgram(cmds).acc;
	},

	B08(contents) {
		const cmds = parseProgram(contents);

		for (let pos = 0; pos < cmds.length; pos++) {
			if (cmds[pos].cmd !== 'acc') {
				const newCmds = cmds.slice();
				newCmds[pos] = {
					...cmds[pos],
					cmd: cmds[pos].cmd === 'jmp' ? 'nop' : 'jmp',
				};

				const exec = execProgram(newCmds);
				if (!exec.loop) {
					return exec.acc;
				}
			}
		}
	},
};

File: lib/08.dart
class Command {
  String cmd;
  int param;
  int count = 0;
}

class Result {
  int acc = 0;
  bool loop = true;
}

List<Command> parseProgram(String contents) {
  return contents.split('\n').map((line) {
    Command cmd = new Command();
    cmd.cmd = line.split(' ').elementAt(0);
    cmd.param = int.parse(line.split(' ').elementAt(1));
    return cmd;
  }).toList();
}

Result execProgram(List<Command> cmds) {
  cmds.forEach((element) {
    element.count = 0;
  });

  Result res = new Result();
  int pos = 0;
  Command curr = cmds.elementAt(pos);
  while(curr.count < 1) {
    curr.count++;
    if (curr.cmd == 'jmp') {
      pos += curr.param;
    } else {
      if(curr.cmd == 'acc') {
        res.acc += curr.param;
      }
      pos++;
    }
    if (cmds.length > pos) {
      curr = cmds.elementAt(pos);
    } else {
      res.loop = false;
      return res;
    }
  }

  return res;
}

int A08(String contents) {
  List<Command> cmds = parseProgram(contents);

  return execProgram(cmds).acc;
}

int B08(String contents) {
  List<Command> cmds = parseProgram(contents);

  for(int pos = 0; pos < cmds.length; pos++) {
    if (cmds[pos].cmd != 'acc') {
      List<Command> newCmds = cmds.toList();
      Command curr = newCmds.elementAt(pos);
      Command newCurr = new Command();
      newCurr.cmd = curr.cmd == 'jmp' ? 'nop' : 'jmp';
      newCurr.param = curr.param;
      newCmds.replaceRange(pos, pos + 1, [newCurr]);

      Result exec = execProgram(newCmds);
      if (!exec.loop) {
        return exec.acc;
      }
    }
  }
}
